File: app/lib/pages/transfer_progress_sheet.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:hyx_app/gen/strings.g.dart';
import 'package:hyx_app/provider/history_provider.dart';
import 'package:hyx_app/provider/transfer_provider.dart';
import 'package:hyx_app/util/formatters.dart';
import 'package:hyx_app/widget/ring_progress.dart';
import 'package:hyx_isolates/rust/api/model.dart' as model;
import 'package:refena_flutter/refena_flutter.dart';

/// 传输进度浮层（BottomSheet）。
///
/// 对应 Kotlin `TransferScreen.TransferringPanel`：环形进度 + 速度 + 已传/总量 +
/// 耗时 + 取消按钮。完成后自动关闭并展示成功/失败 SnackBar。
///
/// 简化设计：作为 showModalBottomSheet 弹出，传输结束后延迟 1.5s 自动关闭，
/// 让用户看到最终状态。
void showTransferProgressSheet(BuildContext context) {
  showModalBottomSheet(
    context: context,
    isScrollControlled: true,
    isDismissible: false,
    enableDrag: false,
    shape: const RoundedRectangleBorder(borderRadius: BorderRadius.vertical(top: Radius.circular(20))),
    builder: (_) => const TransferProgressSheet(),
  );
}

class TransferProgressSheet extends StatefulWidget {
  const TransferProgressSheet({super.key});

  @override
  State<TransferProgressSheet> createState() => _TransferProgressSheetState();
}

class _TransferProgressSheetState extends State<TransferProgressSheet> with Refena {
  bool _autoCloseScheduled = false;

  @override
  Widget build(BuildContext context) {
    final st = context.watch(transferProvider);
    final scheme = Theme.of(context).colorScheme;

    // 终态后安排自动关闭。
    if (st.done && !_autoCloseScheduled) {
      _autoCloseScheduled = true;
      _scheduleAutoClose(st);
    }

    final elapsed = st.startTime != null
        ? DateTime.now().difference(st.startTime!)
        : Duration.zero;
    final eta = etaMillis(st.transferred, st.total, st.speed);

    return Padding(
      padding: EdgeInsets.only(
        left: 24,
        right: 24,
        top: 20,
        bottom: 24 + MediaQuery.of(context).viewInsets.bottom,
      ),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          // 拖把。
          Container(
            width: 40,
            height: 4,
            margin: const EdgeInsets.only(bottom: 16),
            decoration: BoxDecoration(color: scheme.outlineVariant, borderRadius: BorderRadius.circular(2)),
          ),
          // 标题：文件名 + 方向。
          Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Icon(
                st.direction == model.RsTransferDirection.send ? Icons.arrow_upward : Icons.arrow_downward,
                size: 18,
                color: scheme.primary,
              ),
              const SizedBox(width: 6),
              Text(
                st.fileName.isEmpty ? t.transfer.inProgress : st.fileName,
                style: const TextStyle(fontSize: 16, fontWeight: FontWeight.w600),
                maxLines: 1,
                overflow: TextOverflow.ellipsis,
              ),
            ],
          ),
          const SizedBox(height: 20),
          // 环形进度。
          RingProgress(progress: st.fraction, size: 160),
          const SizedBox(height: 20),
          // 状态文本。
          _StatusText(status: st.status, message: st.errorMessage),
          const SizedBox(height: 16),
          // 指标行。
          _MetricRow(
            transferred: st.transferred,
            total: st.total,
            speed: st.speed,
            elapsed: elapsed,
            etaMillis: eta,
          ),
          const SizedBox(height: 20),
          // 取消/完成按钮。
          if (st.busy)
            OutlinedButton.icon(
              onPressed: () => context.redux(transferProvider).dispatch(CancelTransferAction()),
              icon: const Icon(Icons.close),
              label: Text(t.transfer.cancel),
            )
          else
            FilledButton(
              onPressed: () => Navigator.of(context).pop(),
              child: Text(t.transfer.close),
            ),
        ],
      ),
    );
  }

  void _scheduleAutoClose(TransferState st) {
    Future.delayed(const Duration(milliseconds: 1500), () {
      if (!mounted) return;
      // 添加历史记录。
      if (st.startTime != null && st.endTime != null) {
        final record = HistoryRecord(
          id: '${st.startTime!.millisecondsSinceEpoch}',
          direction: st.direction,
          fileName: st.fileName,
          status: st.status,
          bytesTransferred: st.transferred,
          peerAddress: st.peerAddress,
          durationMillis: st.endTime!.difference(st.startTime!).inMilliseconds,
          timestamp: st.endTime!.millisecondsSinceEpoch,
        );
        unawaited(context.redux(historyProvider).dispatchAsync(AddHistoryRecordAction(record)));
      }
      // 展示结果 SnackBar 后关闭。
      final msg = st.status == model.RsTransferStatus.completed
          ? t.transfer.completed
          : st.status == model.RsTransferStatus.failed
              ? '${t.transfer.failed}: ${st.errorMessage ?? ''}'
              : t.transfer.cancelled;
      final ok = st.status == model.RsTransferStatus.completed;
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(
          content: Text(msg),
          backgroundColor: ok ? Colors.green.shade700 : Colors.red.shade700,
          duration: const Duration(seconds: 2),
        ),
      );
      Navigator.of(context).pop();
      // 重置状态，供下次传输。
      context.redux(transferProvider).dispatch(ResetTransferAction());
    });
  }
}

/// 状态文本：根据 status 显示不同颜色。
class _StatusText extends StatelessWidget {
  final model.RsTransferStatus status;
  final String? message;

  const _StatusText({required this.status, this.message});

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final (text, color) = switch (status) {
      model.RsTransferStatus.idle => (t.transfer.idle, scheme.onSurfaceVariant),
      model.RsTransferStatus.pairing => (t.transfer.pairing, scheme.tertiary),
      model.RsTransferStatus.connecting => (t.transfer.connecting, scheme.tertiary),
      model.RsTransferStatus.transferring => (t.transfer.transferring, scheme.primary),
      model.RsTransferStatus.completed => (t.transfer.completed, Colors.green),
      model.RsTransferStatus.failed => ('${t.transfer.failed}${message != null ? ': $message' : ''}', scheme.error),
      model.RsTransferStatus.cancelled => (t.transfer.cancelled, scheme.onSurfaceVariant),
    };
    return Text(
      text,
      style: TextStyle(fontSize: 14, color: color, fontWeight: FontWeight.w500),
      textAlign: TextAlign.center,
      maxLines: 2,
      overflow: TextOverflow.ellipsis,
    );
  }
}

/// 指标行：已传/总量、速度、耗时、剩余。
class _MetricRow extends StatelessWidget {
  final int transferred;
  final int total;
  final double speed;
  final Duration elapsed;
  final int etaMillis;

  const _MetricRow({
    required this.transferred,
    required this.total,
    required this.speed,
    required this.elapsed,
    required this.etaMillis,
  });

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final labelStyle = TextStyle(fontSize: 11, color: scheme.onSurfaceVariant);
    final valueStyle = const TextStyle(fontSize: 14, fontWeight: FontWeight.w600);
    return Row(
      children: [
        _Metric(label: t.transfer.transferred, value: formatBytes(transferred), labelStyle: labelStyle, valueStyle: valueStyle),
        _Metric(label: t.transfer.total, value: formatBytes(total), labelStyle: labelStyle, valueStyle: valueStyle),
        _Metric(label: t.transfer.speed, value: formatSpeed(speed), labelStyle: labelStyle, valueStyle: valueStyle),
        _Metric(label: t.transfer.elapsed, value: formatDuration(elapsed), labelStyle: labelStyle, valueStyle: valueStyle),
        _Metric(
          label: t.transfer.remaining,
          value: etaMillis > 0 ? formatDuration(Duration(milliseconds: etaMillis)) : '—',
          labelStyle: labelStyle,
          valueStyle: valueStyle,
        ),
      ],
    );
  }
}

class _Metric extends StatelessWidget {
  final String label;
  final String value;
  final TextStyle labelStyle;
  final TextStyle valueStyle;

  const _Metric({
    required this.label,
    required this.value,
    required this.labelStyle,
    required this.valueStyle,
  });

  @override
  Widget build(BuildContext context) {
    return Expanded(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          Text(label, style: labelStyle),
          const SizedBox(height: 2),
          Text(value, style: valueStyle, maxLines: 1, overflow: TextOverflow.ellipsis),
        ],
      ),
    );
  }
}

File: android/app/src/main/java/com/ojbkxc/hyx/ui/screens/TransferProgressSheet.kt
package com.ojbkxc.hyx.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.ArrowDownward
import androidx.compose.material.icons.outlined.ArrowUpward
import androidx.compose.material.icons.outlined.Close
import androidx.compose.material3.Button
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip

import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.ojbkxc.hyx.R
import com.ojbkxc.hyx.ui.components.RingProgress
import com.ojbkxc.hyx.ui.components.formatBytes
import com.ojbkxc.hyx.ui.components.formatDuration
import com.ojbkxc.hyx.ui.components.formatSpeed
import com.ojbkxc.hyx.ui.model.TransferDirection
import com.ojbkxc.hyx.ui.model.TransferProgress
import com.ojbkxc.hyx.ui.model.TransferStatus
import com.ojbkxc.hyx.ui.theme.HyxGreen

/**
 * 传输进度浮层（ModalBottomSheet）—— 对齐 Flutter transfer_progress_sheet.dart。
 *
 * 内容：拖把 + 标题(方向箭头+文件名) + 环形进度 + 状态文本 + 5 指标行 + 取消/关闭按钮。
 * 传输中显示 OutlinedButton 取消；终态显示 Button 关闭。
 *
 * @param progress 当前传输进度快照
 * @param status   当前传输状态
 * @param onCancel 取消传输回调
 * @param onDismiss 关闭浮层回调
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun TransferProgressSheet(
    progress: TransferProgress,
    status: TransferStatus,
    onCancel: () -> Unit,
    onDismiss: () -> Unit
) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        dragHandle = null
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(start = 24.dp, end = 24.dp, top = 20.dp, bottom = 24.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            // 1. 拖把。
            Box(
                modifier = Modifier
                    .width(40.dp)
                    .height(4.dp)
                    .clip(RoundedCornerShape(2.dp))
                    .background(MaterialTheme.colorScheme.outlineVariant)
            )
            Spacer(Modifier.height(16.dp))

            // 2. 标题：方向箭头 + 文件名。
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.Center
            ) {
                Icon(
                    imageVector = if (progress.direction == TransferDirection.Send)
                        Icons.Outlined.ArrowUpward else Icons.Outlined.ArrowDownward,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.primary,
                    modifier = Modifier.size(18.dp)
                )
                Spacer(Modifier.size(6.dp))
                Text(
                    if (progress.name.isEmpty()) stringResource(R.string.transfer_in_progress) else progress.name,
                    fontSize = 16.sp,
                    fontWeight = FontWeight.SemiBold,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis
                )
            }
            Spacer(Modifier.height(20.dp))

            // 4. 环形进度。
            RingProgress(progress = progress.fraction, size = 160.dp)
            Spacer(Modifier.height(20.dp))

            // 6. 状态文本。
            StatusText(status = status)
            Spacer(Modifier.height(16.dp))

            // 8. 指标行：已传 / 总量 / 速度 / 已用时 / 剩余。
            MetricRow(progress = progress)
            Spacer(Modifier.height(20.dp))

            // 10. 取消/关闭按钮。
            val busy = status == TransferStatus.Pairing ||
                status == TransferStatus.Connecting ||
                status == TransferStatus.Transferring
            if (busy) {
                OutlinedButton(onClick = onCancel) {
                    Icon(Icons.Outlined.Close, contentDescription = null, modifier = Modifier.size(18.dp))
                    Spacer(Modifier.size(6.dp))
                    Text(stringResource(R.string.cancel))
                }
            } else {
                Button(onClick = onDismiss) {
                    Text(stringResource(R.string.transfer_close))
                }
            }
        }
    }
}

/** 状态文本：根据 status 显示不同颜色。对齐 Flutter _StatusText。 */
@Composable
private fun StatusText(status: TransferStatus) {
    val scheme = MaterialTheme.colorScheme
    val (text, color) = when (status) {
        TransferStatus.Idle -> stringResource(R.string.transfer_idle) to scheme.onSurfaceVariant
        TransferStatus.Pairing -> stringResource(R.string.transfer_pairing) to scheme.tertiary
        TransferStatus.Connecting -> stringResource(R.string.transfer_connecting) to scheme.tertiary
        TransferStatus.Transferring -> stringResource(R.string.transfer_transferring) to scheme.primary
        TransferStatus.Completed -> stringResource(R.string.transfer_completed) to HyxGreen
        TransferStatus.Failed -> stringResource(R.string.transfer_failed) to scheme.error
        TransferStatus.Cancelled -> stringResource(R.string.transfer_cancelled) to scheme.onSurfaceVariant
    }
    Text(
        text,
        fontSize = 14.sp,
        fontWeight = FontWeight.Medium,
        color = color,
        textAlign = TextAlign.Center,
        maxLines = 2,
        overflow = TextOverflow.Ellipsis
    )
}

/** 指标行：已传 / 总量 / 速度 / 已用时 / 剩余。对齐 Flutter _MetricRow。 */
@Composable
private fun MetricRow(progress: TransferProgress) {
    val labelStyle = androidx.compose.ui.text.TextStyle(
        fontSize = 11.sp,
        color = MaterialTheme.colorScheme.onSurfaceVariant
    )
    val valueStyle = androidx.compose.ui.text.TextStyle(
        fontSize = 14.sp,
        fontWeight = FontWeight.SemiBold
    )
    // ETA: 剩余字节 / 速度 * 1000ms。
    val etaMs = if (progress.speedBps > 0 && progress.totalBytes > progress.transferredBytes) {
        ((progress.totalBytes - progress.transferredBytes) / progress.speedBps * 1000).toLong()
    } else 0L

    Row(modifier = Modifier.fillMaxWidth()) {
        Metric(
            label = stringResource(R.string.transfer_transferred),
            value = formatBytes(progress.transferredBytes),
            labelStyle = labelStyle,
            valueStyle = valueStyle
        )
        Metric(
            label = stringResource(R.string.transfer_total),
            value = formatBytes(progress.totalBytes),
            labelStyle = labelStyle,
            valueStyle = valueStyle
        )
        Metric(
            label = stringResource(R.string.transfer_speed),
            value = formatSpeed(progress.speedBps),
            labelStyle = labelStyle,
            valueStyle = valueStyle
        )
        Metric(
            label = stringResource(R.string.transfer_elapsed),
            value = formatDuration(progress.elapsedMs),
            labelStyle = labelStyle,
            valueStyle = valueStyle
        )
        Metric(
            label = stringResource(R.string.transfer_remaining),
            value = if (etaMs > 0) formatDuration(etaMs) else "—",
            labelStyle = labelStyle,
            valueStyle = valueStyle
        )
    }
}

/** 单个指标列（等宽）。对齐 Flutter _Metric。 */
@Composable
private fun Metric(
    label: String,
    value: String,
    labelStyle: androidx.compose.ui.text.TextStyle,
    valueStyle: androidx.compose.ui.text.TextStyle
) {
    Column(
        modifier = Modifier.weight(1f),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text(label, style = labelStyle)
        Spacer(Modifier.height(2.dp))
        Text(value, style = valueStyle, maxLines = 1, overflow = TextOverflow.Ellipsis)
    }
}